### ISSUE TITLE
failed installed with i18n.gettext are not removed on uninstall

### ISSUE THREAD

**Issue by keszybz** [mesonbuild/meson#1601]
After `ninja-build install`, `ninja-build uninstall` I see 
```
$ find /var/tmp/inst7 -type f
/var/tmp/inst7/usr/share/locale/be/LC_MESSAGES/systemd.mo
/var/tmp/inst7/usr/share/locale/be@latin/LC_MESSAGES/systemd.mo
...
/var/tmp/inst7/usr/share/locale/zh_TW/LC_MESSAGES/systemd.mo
```
All other installed files seem to be removed, so it seems only `i18n` does not support that.

**Comment by inigomartinez** [comment]
Files installed by `gnome.yelp` aren't also installed. Neither files installed by `i18n.gettext` nor `gnome.yelp` are present in the `meson-logs/install-log.txt` file.

**Comment by TingPing** [comment]
Yes any custom installation provided by modules is not currently tracked, we need to add some helper function to do so.

**Comment by raveit65** [comment]
I like to push the report again.
We see that problem with several packages, currently we are switching MATE desktop to use meson build system.
https://github.com/mate-desktop/mate-backgrounds/pull/22#issuecomment-489966620
and
https://github.com/mate-desktop/mate-desktop/pull/333#issuecomment-483708111
Would be nice if this can be fixed.
Thanks.

**Comment by ericonr** [comment]
@TingPing what sort of help is necessary for this to happen? I can try to add the feature.

**Comment by dcbaker** [comment]
There's several lists in the `Build` class (`mesonbuild/build.py`), which is in the `ModuleState`, `man`, `data`, and `headers` which modules need to update. There's code in the `mesonbuild/interpreter.py` that does a bunch of the front end parsing and updates these. I'm not sure exactly what it would look like, but that's where I'd start.

**Comment by andy5995** [comment]
Anyone have suggestions as to user-friendly ways to uninstall the mo files? I was thinking of generating an `uninstall_langs` shell script, based on an `.in` file that contains variables which get replaced by the paths to the mo files.

**Comment by andy5995** [comment]
> Anyone have suggestions as to user-friendly ways to uninstall the mo files? I was thinking of generating an `uninstall_langs` shell script, based on an `.in` file that contains variables which get replaced by the paths to the mo files.

```meson
if get_option('nls')
  ul = configuration_data()
  ul.set('LOCALEDIR', localedir)
  ul.set('MESON_SOURCE_ROOT', meson.source_root())
  ul.set('MESON_PROJECT_NAME', meson.project_name())
  configure_file(input : 'uninstall_langs.sh.in',
                 output : 'uninstall_langs.sh',
                 configuration : ul)
endif
```

```sh
#!/bin/sh

LINGUAS=$(cat @MESON_SOURCE_ROOT@/po/LINGUAS)

for lang in $LINGUAS; do
  file="@LOCALEDIR@/$lang/LC_MESSAGES/@MESON_PROJECT_NAME@.mo"
  if test -f "$file"; then
    echo "removing $file"
    /bin/rm "$file"
  else
    echo "$file does not exist"
  fi
done
```

**Comment by Volker-Weissmann** [comment]
> Some maintainer might want to reopen this issue or create a similar one for gnome.yelp and any other module. Those still need to be fixed.

26c1869a142a952ffa23fe566a255b259304a39b fixed this for gnome.yelp.
